geopy/geocoders/baidu.py

```diff
@@ -205,44 +224,55 @@ def _check_status(status):
         """
         Validates error statuses.
         """
-        if status == '0':
+        if status == 0:
             # When there are no results, just return.
             return
-        if status == '1':
+        if status == 1:
             raise GeocoderQueryError(
                 'Internal server error.'
             )
-        elif status == '2':
+        elif status == 2:
             raise GeocoderQueryError(
                 'Invalid request.'
             )
-        elif status == '3':
+        elif status == 3:
             raise GeocoderAuthenticationFailure(
                 'Authentication failure.'
             )
-        elif status == '4':
+        elif status == 4:
             raise GeocoderQuotaExceeded(
                 'Quota validate failure.'
             )
-        elif status == '5':
+        elif status == 5:
             raise GeocoderQueryError(
                 'AK Illegal or Not Exist.'
             )
-        elif status == '101':
+        elif status == 101:
             raise GeocoderQueryError(
                 'Your request was denied.'
             )
-        elif status == '102':
+        elif status == 102:
             raise GeocoderQueryError(
                 'IP/SN/SCODE/REFERER Illegal:'
             )
-        elif status == '2xx':
+        elif status in range(200, 300):
             raise GeocoderQueryError(
                 'Has No Privilleges.'
             )
-        elif status == '3xx':
+        elif status in range(300, 400):
             raise GeocoderQuotaExceeded(
                 'Quota Error.'
             )
         else:
             raise GeocoderQueryError('Unknown error')
+
+    def _url_with_signature(self, quoted_params):
+        """
+        Return the request url with signature.
+        qutoted_params is the escaped str of params in request.
+        """
+
+        # Since quoted_params is escaped by urlencode, don't apply quote twice
+        raw = self.api_path + '?' + quoted_params + self.security_key
```

Comment:
According to the docs at http://lbsyun.baidu.com/index.php?title=lbscloud/api/appendix , the `quote(params, safe="/:=&?#+!$,;'@()*[]")` code is used for getting a raw string, while there we have `urlencode(params)`. Wouldn't that cause different hashes for peculiar requests? I think the relevant test cases should be added. For example, a one with a comma: `locator.geocode("beijing, china")`.

Reply:
The `quoted_params` in `self._url_with_signature` was given by `urlencode`. `urlencode` has already calls `quote_plus` on parameters internally. Except parameters, `quote` never escape the characters in site/path/key parts, because they are not special characters.

Reply:
Yes, this patch code doesn't match the given sample code in docs. But it works.

Reply:
It sure works, but for some cases it might not work. Could you please add the test with a comma, and fix the code, if that fails?

Reply:
It's better to stick to docs. I'm going to  change `urlencode(params)` to `urlencode(params, safe="!*'();:@&=+$,/?#[]", quote_via=quote)`.  Will you accept this change?

Reply:
Please add the test first :)

But yeah, that would be better than crafting the query string manually. Even better if all of this is not required and a simple `urlencode` works.